src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -344,17 +345,35 @@ public Project AddProject(string name, string assemblyName, string language)
     public Solution AddProject(ProjectId projectId, string name, string assemblyName, string language)
         => this.AddProject(ProjectInfo.Create(projectId, VersionStamp.Create(), name, assemblyName, language));
 
-    /// <summary>
-    /// Create a new solution instance that includes a project with the specified project information.
-    /// </summary>
+    /// <inheritdoc cref="SolutionCompilationState.AddProjects"/>
     public Solution AddProject(ProjectInfo projectInfo)
-        => WithCompilationState(_compilationState.AddProject(projectInfo));
+    {
+        using var _ = ArrayBuilder<ProjectInfo>.GetInstance(1, out var projectInfos);
+        projectInfos.Add(projectInfo);
+        return AddProjects(projectInfos);
+    }
 
-    /// <summary>
-    /// Create a new solution instance without the project specified.
-    /// </summary>
+    /// <inheritdoc cref="SolutionCompilationState.AddProjects"/>
+    internal Solution AddProjects(ArrayBuilder<ProjectInfo> projectInfos)
+    {
+        Contract.ThrowIfTrue(projectInfos.Select(i => i.Id).Distinct().Count() != projectInfos.Count, "Duplicate ProjectId provided");
+        return WithCompilationState(_compilationState.AddProjects(projectInfos));
+    }
+
+    /// <inheritdoc cref="SolutionCompilationState.RemoveProjects"/>
     public Solution RemoveProject(ProjectId projectId)
-        => WithCompilationState(_compilationState.RemoveProject(projectId));
+    {
+        using var _ = ArrayBuilder<ProjectId>.GetInstance(1, out var projectIds);
+        projectIds.Add(projectId);
+        return RemoveProjects(projectIds);
+    }
+
+    /// <inheritdoc cref="SolutionCompilationState.RemoveProjects"/>
+    internal Solution RemoveProjects(ArrayBuilder<ProjectId> projectIds)
+    {
+        Contract.ThrowIfTrue(projectIds.Distinct().Count() != projectIds.Count, "Duplicate ProjectId provided");
```

Comment:
>projectIds.Distinct().Count() != projectIds.Count [](http://example.com/codeflow?start=29&length=49)

Might be nice if we had a HasDuplicates method for ArrayBuilder similar to wha we have for ImmutableArray

Reply:
done.

Reply:
Just wanted to make sure this is at the right level, intead of down in SolutionCompilationState or even SolutionState

Reply:
This is our public entrypoint for mutation, and most of the rest of hte other entrypoints do checks here as well.  That said, some checks are pushed inwards.  I can't really determine the "why" of that.  i'll push in deeper though just to ensure that we don't accidentally bypass this.